File: golang/kalman.go
package main

import (
	"math"

	"gonum.org/v1/gonum/stat/distuv"
)

type Kalman struct {
	Battery *Battery
	SoC_OCV SoCOCV
	Pk      [][]float64 // Error Covariance
	Xk      [][]float64 // State Estimate
	Kk      [][]float64 // Kalman Gain
	Fk      [][]float64 // Process Jacobian
	Bk      [][]float64 // Control Input Matrix
	SigmaWk [][]float64 // Process Noise Covariance Matrix
	SigmaVk float64     // Variance of the voltage sensor
	Yk      float64     // Estimated Terminal Voltage
}

// Initialize a new Kalman Filter for a specific battery and
// SOC-OCV Curve data.
func NewKalman(b *Battery, d SoCOCV) Kalman {
	var k Kalman

	k.Battery = b
	k.SoC_OCV = d

	// Initial State
	k.Xk = MatT([][]float64{{b.Zk, 0, 0}})
	// Initial Error Covariance
	k.Pk = [][]float64{
		{0.005, 0, 0},
		{0, 0.0001, 0},
		{0, 0, 0.0001},
	}
	// Process Jacobian
	k.Fk = [][]float64{
		{1, 0, 0},
		{0, math.Exp(-k.Battery.Dt / (k.Battery.R1 * k.Battery.C1)), 0},
		{0, 0, math.Exp(-k.Battery.Dt / (k.Battery.R2 * k.Battery.C2))},
	}

	// Control Input Matrix
	k.Bk = [][]float64{
		{-k.Battery.Ni * k.Battery.Dt / (3600 * k.Battery.Cn)},
		{1 - math.Exp(-k.Battery.Dt/(k.Battery.R1*k.Battery.C1))},
		{1 - math.Exp(-k.Battery.Dt/(k.Battery.R2*k.Battery.C2))},
	}

	k.SigmaWk = MatMul(k.Bk, MatT(k.Bk))
	k.SigmaVk = math.Pow(0.035, 2)

	return k
}

// Hk is the Measurement Jacobian
func (k *Kalman) GetHk() [][]float64 {
	// the rate of change of the Open Circuit Voltage to the State of Charge
	derivative := k.SoC_OCV.Derivative(k.Xk[0][0] * 100)

	Hk := [][]float64{
		{derivative, -k.Battery.R1, -k.Battery.R2},
	}

	return Hk
}

// StepOne: State Estimate Propagation
// In this step we propagate the previous state to the current state
// using the input current to predict the current state.
// X_{k+1} = F_k X_k + B_k i_k + W_k
func (k *Kalman) StepOne() {
	val := MatMulC(k.Bk, k.Battery.I)
	k.Xk = MatAdd(MatMul(k.Fk, k.Xk), val)
}

// StepTwo: Error Covariance Propagation
// In this step we propagate the error covariance of the system.
// Error covariance is used to count for errors in the measurement
// and in prediction. The Error Covariance uses a gaussian distribution
// for error margains rather than the completely random distribution
func (k *Kalman) StepTwo() {
	k.Pk = MatAdd(MatMul(MatMul(k.Fk, k.Pk), MatT(k.Fk)), k.SigmaWk)
}

// StepThree: Predict System Output
// The system uses 2nd-order Thevenin equivalent circuit to simulate the
// Car Battery, we calculate the terminal voltage using the Battery parameters
// and get the OCV using the SOC-OCV graph using the current SoC.
func (k *Kalman) StepThree() {
	OCV := k.SoC_OCV.GetVoltage(k.Xk[0][0] * 100)
	k.Yk = OCV - k.Battery.R0*k.Battery.I - k.Battery.R1*k.Battery.I1 - k.Battery.R2*k.Battery.I2
}

// StepFour: Calculate the kalman gain
// Kalman gain is used as an indication of the stability of the system.
// In the first iterations, the measured values of voltage are considered
// to be more accurate and though has more weight than the predicted output.
// as the system matures, the measured voltage weight becomes lesser as the
// system becomes more stable and based on many measurements.
func (k *Kalman) StepFour() {
	hk := k.GetHk()

	HKxPKxHKt := MatMul(MatMul(hk, k.Pk), MatT(hk))
	SigmaY := MatDiffuse(HKxPKxHKt) + k.SigmaVk

	k.Kk = MatMulC(MatMul(k.Pk, MatT(hk)), 1/SigmaY)
}

// StepFive: State Esitmate Update
// We update the system state using the measured and predicted terminal
// voltages. The severity of the update is based on the kalman gain
// calculated in the previous step.
func (k *Kalman) StepFive(measuredVoltage float64) {
	k.Xk = MatAdd(k.Xk, MatMulC(k.Kk, k.Yk-measuredVoltage))
}

// StepSix: Error Covariance Update
// We update the error covariance to reflect the maturity of the system
// as more measured values are added to the system.
func (k *Kalman) StepSix() {
	k.Pk = MatMul(MatSub(MatI(3), MatMul(k.Kk, k.GetHk())), k.Pk)
}

// Cycle: Represents a full Kalman Filter Cycle
// measuredVoltage: the terminal voltage of the battery.
// measuredCurrent: The current drawn from the battery to feed the system.
func (k *Kalman) Cycle(measuredCurrent, measuredVoltage float64) float64 {
	k.Battery.UpdateCurrents(measuredCurrent)

	// Coulombic Efficiency
	// 1  at discharge
	// Ni at charge
	Ni := k.Battery.Ni
	if measuredCurrent > 0 {
		Ni = 1
	}

	k.Bk[0][0] = -Ni * k.Battery.Dt / (3600 * k.Battery.Cn)

	// prediction
	k.StepOne()
	k.StepTwo()
	k.StepThree()

	// update
	k.StepFour()
	k.StepFive(measuredVoltage)
	k.StepSix()

	return k.Xk[0][0]
}

// MockCycle: full Kalman Filter Cycle, this is used for demonstration
// purposes. In this MockCycle the measuredVoltage equals to the
// predicted voltage + gaussian noise. it aims to simulate the working
// of a voltage sensor
func (k *Kalman) MockCycle(I float64) (float64, float64) {
	k.Battery.UpdateCurrents(I)

	// Coulombic Efficiency
	Ni := k.Battery.Ni
	if I > 0 {
		Ni = 1
	}

	k.Bk[0][0] = -Ni * k.Battery.Dt / (3600 * k.Battery.Cn)

	// prediction
	k.StepOne()
	k.StepTwo()
	k.StepThree()

	// update
	mockV := addGaussianError(k.Yk, k.SigmaVk)
	k.StepFour()
	k.StepFive(mockV)
	k.StepSix()

	return k.Xk[0][0], k.Yk
}

func addGaussianError(mean, sigma float64) float64 {
	normalDist := distuv.Normal{
		Mu:    mean,
		Sigma: sigma,
	}

	return normalDist.Rand()
}

File: c/kalman.c
#include "kalman.h"
#include <math.h>


// Helper Functions
float* allocate_1(float element);
float* allocate_3(float e1, float e2, float e3);
float **allocate_3_ptrs(float *ptr1, float *ptr2, float *ptr3);
float **allocate_1_ptr(float *ptr);

void StepOne(Kalman *k, float measuredCurrent) {
	float **BkxI = MatMulC(k->Bk, 3, 1, measuredCurrent);
	float **FkxXk = MatMul(k->Fk, k->Xk, 3, 3, 3, 1);

	freeMatrix(k->Xk, 3);

	k->Xk = MatAdd(FkxXk, BkxI, 3, 1);

	freeMatrix(BkxI, 3);
	freeMatrix(FkxXk, 3);
}

void StepTwo(Kalman *k) {
	float **FkxPk = MatMul(k->Fk, k->Pk, 3, 3, 3, 3);
	float **Fkt = MatT(k->Fk, 3, 3);
	float **FkxPkxFkt = MatMul(FkxPk, Fkt, 3, 3, 3, 3);

	freeMatrix(k->Pk, 3);

	k->Pk = MatAdd(FkxPkxFkt, k->SigmaWk, 3, 3);

	freeMatrix(FkxPkxFkt, 3);
	freeMatrix(Fkt, 3);
	freeMatrix(FkxPk, 3);
}

void StepThree(Kalman *k, float measuredCurrent) {
	// get the closest reading to the SoC.
	float OCV = Get_Voltage(k->Xk[0][0] * 100);
	k->Yk = OCV - k->b->R0 * measuredCurrent + k->Hk[0][1] * k->Xk[1][0] + k->Hk[0][2] * k->Xk[2][0];
}

void StepFour(Kalman *k) {
	float **HkxPk = MatMul(k->Hk, k->Pk, 1, 3, 3, 3);
	float **Hkt = MatT(k->Hk, 1, 3);
	float **HkxPkxHkt = MatMul(HkxPk, Hkt, 1, 3, 3, 1);

	float SigmaY = HkxPkxHkt[0][0] + k->SigmaVk;

	float **PkxHkt = MatMul(k->Pk, Hkt, 3, 3, 3, 1);

	freeMatrix(k->Kk, 3);

	k->Kk = MatMulC(PkxHkt, 3, 1, SigmaY);

	freeMatrix(HkxPk, 1);
	freeMatrix(Hkt, 3);
	freeMatrix(HkxPkxHkt, 1);
	freeMatrix(PkxHkt, 3);
}

void StepFive(Kalman *k, float measuredVoltage) {
	float **KkxVolt = MatMulC(k->Kk, 3, 1, k->Yk - measuredVoltage);
	float **oldXk = k->Xk;

	k->Xk = MatAdd(k->Xk, KkxVolt, 3, 1);

	freeMatrix(KkxVolt, 3);
	freeMatrix(oldXk, 3);
}

void StepSix(Kalman *k) {
	float **I = MatI(3);
	float **KkxHk = MatMul(k->Kk, k->Hk, 3, 1, 1, 3);
	float **I_KkxHk = MatSub(I, KkxHk, 3, 3);
	float **oldPk = k->Pk;

	k->Pk = MatMul(I_KkxHk, k->Pk, 3, 3, 3, 3);

	freeMatrix(I, 3);
	freeMatrix(KkxHk, 3);
	freeMatrix(I_KkxHk, 3);
	freeMatrix(oldPk, 3);
}


Kalman* InitKalman(Battery* b) {
	Kalman *k = NULL;
	k = (Kalman *)malloc(sizeof(Kalman));

	k->b = b;

	// Initial State
	k->Xk = allocate_3_ptrs(
		allocate_1(b->Zk),
		allocate_1(0),
		allocate_1(0)
	);

	// Initial Error Covariance
	k->Pk = allocate_3_ptrs(
		allocate_3(0.005, 0, 0),
		allocate_3(0, 0.0001, 0),
		allocate_3(0, 0, 0.001)
	);

	// Process Jacobian
	k->Fk = allocate_3_ptrs(
		allocate_3(1, 0, 0),
		allocate_3(0,expf(-k->b->Dt / (k->b->R1 * k->b->C1)) ,0),
		allocate_3(0,0,expf(-k->b->Dt / (k->b->R2 * k->b->C2)))
	);


	k->Bk = allocate_3_ptrs(
		allocate_1(-k->b->Ni * k->b->Dt / (3600 * k->b->Cn)),
		allocate_1(1 - expf(-k->b->Dt / (k->b->R1 * k->b->C1))),
		allocate_1(1 - expf(-k->b->Dt / (k->b->R2 * k->b->C2)))
	);


	float derivative = Get_Derivative(k->Xk[0][0] * 100);

	k->Hk = allocate_1_ptr(allocate_3(derivative, -k->b->R1, -k->b->R2));

	float **Bkt = MatT(k->Bk, 3, 1);
	k->SigmaWk = MatMul(k->Bk, Bkt, 3, 1, 1, 3);
	k->SigmaVk = pow(0.035, 2);

	freeMatrix(Bkt, 1);
	return k;
}



SoC_Reading KalmanCycle(Kalman* k, float measuredCurrent, float measuredVoltage) {
	float Ni = k->b->Ni;
	if (measuredCurrent > 0) {
		Ni = 1;
	}
	
	k->Bk[0][0] = - Ni * k->b->Dt / (3600 * k->b->Cn);

	StepOne(k, measuredCurrent);
	StepTwo(k);
	StepThree(k, measuredCurrent);

	k->Hk[0][0] = Get_Derivative(k->Xk[0][0] * 100);

	StepFour(k);
	StepFive(k, measuredVoltage);
	StepSix(k);

	return k->Xk[0][0];
}


SoC_Reading KalmanMockCycle(Kalman* k, float measuredCurrent, float *voltage) {
	float Ni = k->b->Ni;
	if (measuredCurrent > 0) {
		Ni = 1;
	}
	
	k->Bk[0][0] = - Ni * k->b->Dt / (3600 * k->b->Cn);

	StepOne(k, measuredCurrent);
	StepTwo(k);
	StepThree(k, measuredCurrent);

	k->Hk[0][0] = Get_Derivative(k->Xk[0][0] * 100);

	StepFour(k);
	StepFive(k, k->Yk);
	StepSix(k);

	*voltage = k->Yk;
	return k->Xk[0][0];
}


float* allocate_3(float e1, float e2, float e3) {
	float *arr = NULL;

	arr = (float *)malloc(sizeof(float) * 3);
	if (arr == NULL) {
		return arr;
	}

	arr[0] = e1;
	arr[1] = e2;
	arr[2] = e3;

	return arr;
}

float* allocate_1(float element) {
	float *arr = NULL;

	arr = (float *)malloc(sizeof(float));
	if (arr == NULL) {
		return arr;
	}

	arr[0] = element;

	return arr;
}

float **allocate_3_ptrs(float *ptr1, float *ptr2, float *ptr3) {
	float **arr = NULL;

	arr = (float **)malloc(sizeof(float *) * 3);
	if (arr == NULL) {
		return arr;
	}
	
	arr[0] = ptr1;
	arr[1] = ptr2;
	arr[2] = ptr3;

	return arr;
}

float **allocate_1_ptr(float *ptr) {
	float **arr = NULL;

	arr = (float **)malloc(sizeof(float *));
	if (arr == NULL) {
		return arr;
	}
	
	arr[0] = ptr;

	return arr;
}
